File: rust/parity/specs/rust_solana_cpi_instruction_sentry.py
"""Parity spec for PiRustSolanaCPIInstructionSentry.

A spec declares:
  RUST_NAME       - registry key passed to pi_core.run_agent
  run_python(data)-> dict  - run the ORIGINAL python agent, return model_dump()
  SAMPLES         - list of {"input": {...}, "env": {...optional...}}
"""
from _util import load_py_agent

RUST_NAME = "PiRustSolanaCPIInstructionSentry"

_mod = load_py_agent("pi_rust_solana_cpi_instruction_sentry.py")


def run_python(data: dict) -> dict:
    agent = _mod.PiRustSolanaCPIInstructionSentry()
    out = agent.audit_cpi_instruction(_mod.SolanaCPIInstructionInput(**data))
    return out.model_dump()


# No CPI at all -> secure / PASSED.
_CLEAN = "fn process(ctx: Context) { msg!(\"hello\"); }"

# invoke() with no key/id/check token in the body -> flagged.
_VULN_INVOKE = "fn transfer(ctx: Context) { invoke(&ix, accounts); }"

# invoke_signed() with no key/id/check token -> flagged (substring 'invoke').
_VULN_SIGNED = "fn mint(ctx: Context) { invoke_signed(&ix, accounts, seeds); }"

# invoke() but body mentions '.key' -> guard satisfied, secure.
_SAFE_KEY = "fn transfer(ctx: Context) { assert_eq!(prog.key, expected); invoke(&ix, accs); }"

# invoke() but body mentions 'program_id' (contains 'id') -> secure.
_SAFE_ID = "fn transfer(ctx: Context) { require(prog == program_id); invoke(&ix, accs); }"

# invoke() but body mentions 'check' -> secure.
_SAFE_CHECK = "fn transfer(ctx: Context) { check_program(prog); invoke(&ix, accs); }"

# Multiple methods across newlines: one safe, one vulnerable.
_MULTI = "\n".join(
    [
        "fn safe_one(ctx: Context) {",
        "    assert_eq!(prog.key, expected);",
        "    invoke(&ix, accs);",
        "}",
        "",
        "fn bad_one(ctx: Context) {",
        "    invoke(&ix, accs);",
        "}",
    ]
)

SAMPLES = [
    {"input": {"file_path": "lib.rs", "rust_code": _CLEAN}},
    {"input": {"file_path": "lib.rs", "rust_code": _VULN_INVOKE}},
    {"input": {"file_path": "lib.rs", "rust_code": _VULN_SIGNED}},
    {"input": {"file_path": "lib.rs", "rust_code": _SAFE_KEY}},
    {"input": {"file_path": "lib.rs", "rust_code": _SAFE_ID}},
    {"input": {"file_path": "lib.rs", "rust_code": _SAFE_CHECK}},
    {"input": {"file_path": "lib.rs", "rust_code": _MULTI}},
    {"input": {"file_path": "lib.rs", "rust_code": ""}},
    {"input": {"file_path": "lib.rs", "rust_code": _VULN_INVOKE, "check_level": "MEDIUM"}},
    # non-strict env -> WARN path, is_secure coerced back to True
    {"input": {"file_path": "lib.rs", "rust_code": _VULN_INVOKE},
     "env": {"PI_SOLANA_CPI_INSTRUCTION_STRICT_MODE": "false"}},
    {"input": {"file_path": "lib.rs", "rust_code": _VULN_INVOKE},
     "env": {"PI_SOLANA_CPI_INSTRUCTION_STRICT_MODE": "true"}},
]

File: rust/crates/pi-agents/src/agents/rust_solana_cpi_instruction_sentry.rs
//! Port of `pi_micro_agents/pi_rust_solana_cpi_instruction_sentry.py`.
//!
//! Specialized Rust/Solana micro-agent that audits Solana smart contracts for
//! secure CPI (Cross-Program Invocation) program validation. Behaviour is a
//! line-for-line mirror of the Python original.

use once_cell::sync::Lazy;
use regex::Regex;
use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize)]
pub struct Input {
    pub file_path: String,
    pub rust_code: String,
    #[serde(default = "default_check_level")]
    pub check_level: String,
}

fn default_check_level() -> String {
    "STRICT".to_string()
}

#[derive(Debug, Serialize, PartialEq)]
pub struct Output {
    pub is_secure: bool,
    pub vulnerable_elements: Vec<String>,
    pub flagged_findings: Vec<String>,
    pub risk_score: f64,
    pub status: String,
}

/// Mirrors `is_strict_mode()`: strict unless the env var is set to a value that
/// is not (case-insensitively) "true".
fn is_strict_mode() -> bool {
    match std::env::var("PI_SOLANA_CPI_INSTRUCTION_STRICT_MODE") {
        Ok(v) => v.to_lowercase() == "true",
        Err(_) => true,
    }
}

// Mirrors Python `re.findall(r'fn\s+([a-zA-Z0-9_]+)\s*\((.*?)\)[^{]*\{([\s\S]*?)\}', code)`.
// 3 capture groups. The Python pattern has no DOTALL flag, so `.` (in the args
// group) does not match newlines, while `[\s\S]` (in the body group) matches
// every character including newlines. The Rust `regex` crate matches both of
// those behaviours by default.
static METHOD_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"fn\s+([a-zA-Z0-9_]+)\s*\((.*?)\)[^{]*\{([\s\S]*?)\}").unwrap()
});

pub fn audit_cpi_instruction(input: &Input) -> Output {
    let code = &input.rust_code;
    let mut vulnerable_elements: Vec<String> = Vec::new();
    let mut flagged_findings: Vec<String> = Vec::new();

    for caps in METHOD_RE.captures_iter(code) {
        // group 1 = name, group 2 = args, group 3 = body
        let name = caps.get(1).map(|m| m.as_str()).unwrap_or("");
        let _args = caps.get(2).map(|m| m.as_str()).unwrap_or("");
        let body = caps.get(3).map(|m| m.as_str()).unwrap_or("");

        if body.contains("invoke") || body.contains("invoke_signed") {
            // CPI occurs. Check if the target program id is checked or validated.
            if !body.contains("key") && !body.contains("id") && !body.contains("check") {
                vulnerable_elements.push(name.to_string());
                flagged_findings.push(format!(
                    "Instruction handler '{name}' invokes CPI but does not explicitly validate the target program account ID. \
Malicious actors could pass a spoofed program ID to intercept CPI parameters or return spoofed results."
                ));
            }
        }
    }

    let mut is_secure = vulnerable_elements.is_empty();
    let risk_score = if !is_secure { 75.0 } else { 0.0 };

    let is_strict = is_strict_mode();
    let mut status = "PASSED".to_string();
    if !is_secure {
        if is_strict {
            status = "REJECTED_SOLANA_CPI_INSTRUCTION".to_string();
        } else {
            status = "WARN_SOLANA_CPI_INSTRUCTION".to_string();
            is_secure = true;
        }
    }

    Output {
        is_secure,
        vulnerable_elements,
        flagged_findings,
        risk_score,
        status,
    }
}

pub fn run_json(input_json: &str) -> Result<String, String> {
    let input: Input = serde_json::from_str(input_json).map_err(|e| e.to_string())?;
    let out = audit_cpi_instruction(&input);
    serde_json::to_string(&out).map_err(|e| e.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn run(code: &str) -> Output {
        audit_cpi_instruction(&Input {
            file_path: "lib.rs".into(),
            rust_code: code.into(),
            check_level: "STRICT".into(),
        })
    }

    #[test]
    fn secure_code_passes() {
        // No CPI invocation at all -> secure.
        let o = run("fn process(ctx: Context) { msg!(\"hi\"); }");
        assert!(o.is_secure);
        assert_eq!(o.status, "PASSED");
        assert_eq!(o.risk_score, 0.0);
        assert!(o.vulnerable_elements.is_empty());
    }

    #[test]
    fn invoke_without_validation_flagged() {
        let o = run("fn transfer(ctx: Context) { invoke(&ix, accounts); }");
        assert!(!o.is_secure);
        assert_eq!(o.status, "REJECTED_SOLANA_CPI_INSTRUCTION");
        assert_eq!(o.risk_score, 75.0);
        assert_eq!(o.vulnerable_elements, vec!["transfer"]);
    }

    #[test]
    fn invoke_with_key_check_is_secure() {
        // body contains "key" so the validation guard is satisfied.
        let o = run("fn transfer(ctx: Context) { assert_eq!(prog.key, expected); invoke(&ix, accounts); }");
        assert!(o.is_secure);
        assert_eq!(o.status, "PASSED");
        assert!(o.vulnerable_elements.is_empty());
    }
}
